src/Build/Evaluation/Expander.cs

```diff
@@ -4677,6 +4694,143 @@ private static bool TryGetArgs(object[] args, out string arg0, out int arg1)
                 return false;
             }
 
+            private static bool IsFloatingPointRepresentation(object value)
+            {
+                const char numberDecimalSeparator = '.';
+
+                return value is double || (value is string str && str.Contains(numberDecimalSeparator));
+            }
+
+            private static bool TryExecuteAdd(object[] args, out object resultValue)
+            {
+                resultValue = null;
+
+                if (args.Length != 2)
+                {
+                    return false;
+                }
+
+                if (!IsIntrinsicFunctionOverloadsEnabled() || IsFloatingPointRepresentation(args[0]) || IsFloatingPointRepresentation(args[1]))
+                {
+                    if (TryConvertToDouble(args[0], out double arg0) && TryConvertToDouble(args[1], out double arg1))
+                    {
+                        resultValue = IntrinsicFunctions.Add(arg0, arg1);
+                        return true;
+                    }
+                }
+                else if (TryConvertToLong(args[0], out long arg0) && TryConvertToLong(args[1], out long arg1))
+                {
+                    resultValue = IntrinsicFunctions.Add(arg0, arg1);
+                    return true;
+                }
+
+                return false;
```

Comment:
I have two asks for this code:

1) can we extract this to prevent repetition?

2) This code doesn't grace handle large integers (no decimals, valua above long max) - e.g. "12345678901234567890123456". How about first trying to convert to long, and fallback to double if that is not possible?

Reply:
There is a lot of repetition in the handling of the intrinsic functions and there is an idea that has been floated that there should be a code generator for it.

Yes, the `TryExecute[Add|Subtract|Multiply|Divide|Modulo]` methods don't handle large integers and they don't handle mixed arguments of `long` and `double` but both cases are handled elsewhere in the code.

The methods could be changed to extend the `if ... else if ...` to try `double` when `long` fails. Simplifying the code to try `long` first and falling back to `double` imposes a cost on using a `double`. The `IsFloatingPointRepresentation()` method is intended to be a quick test without attempting a full parse.

Reply:
Thanks for the current extraction of the common code!
As for handling large integers - this is fine now as well, I'd just like to see an explicit unit test for that (to prevent future breaks)